Dockerfile

```diff
@@ -1,18 +1,66 @@
-FROM python:3.8-alpine
+FROM python:3.9.6-alpine3.14 as builder
 LABEL maintainer="Jacob Tomlinson <jacob@tomlinson.email>"
+LABEL maintainer="Rémy Greinhofer <remy.greinhofer@gmail.com>"
 
 WORKDIR /usr/src/app
-ARG EXTRAS=.[all,connector_matrix_e2e]
+
+ARG EXTRAS=[all,connector_matrix_e2e]
+ENV DEPS_DIR=/usr/src/app/deps
 
 # Copy source
 COPY . .
 
+# Install build tools and libraries to build OpsDroid and its dependencies.
 RUN apk update \
-    && apk add --no-cache build-base linux-headers musl-dev python3-dev libzmq zeromq-dev git openssh-client olm olm-dev libffi-dev openssl-dev cargo \
-    && pip3 install --upgrade pip \
-    && pip3 install --no-cache-dir $EXTRAS \
-    && apk del build-base linux-headers musl-dev python3-dev zeromq-dev
+    && apk add \
+    build-base \
+    cargo \
+    gcc \
+    git \
+    g++ \
+    libffi-dev \
+    linux-headers \
+    musl-dev \
+    olm-dev \
+    openssh-client \
+    openssl-dev \
+    python3-dev \
+    zeromq-dev \
+    && pip install --upgrade \
+    build \
+    pip \
+    setuptools \
+    setuptools-scm \
+    wheel \
+    && mkdir -p "${DEPS_DIR}" \
+    && pip download --use-feature=in-tree-build --prefer-binary -d ${DEPS_DIR} .${EXTRAS} \
+    && pip wheel -w ${DEPS_DIR} ${DEPS_DIR}/*.tar.gz \
+    && python -m build --wheel --outdir ${DEPS_DIR}
+
+FROM python:3.9.6-alpine3.14 as runtime
+LABEL maintainer="Jacob Tomlinson <jacob@tomlinson.email>"
+LABEL maintainer="Rémy Greinhofer <remy.greinhofer@gmail.com>"
+
+WORKDIR /usr/src/app
+
+ARG EXTRAS=[all,connector_matrix_e2e]
+ENV DEPS_DIR=/usr/src/app/deps
+
+# Copy the pre-built dependencies.
+COPY --from=builder ${DEPS_DIR}/*.whl ${DEPS_DIR}/
+
+# Install Opsdroid using only pre-built dependencies.
+RUN apk add --no-cache \
+    git \
```

Comment:
As I said on the other PR I am pretty sure we need olm at runtime here.

Reply:
Done :+1: .